source/toolbox/memory/Array.h

```diff
@@ -43,6 +44,72 @@ template <class TYPE> class Pointer;
 template <class TYPE>
 class Array
 {
+private:
+
+   class Allocator
+   {
+   private:
+      constexpr static size_t get_block_id(const size_t block_size) {
+         return (block_size < 2 ? 0 : std::ilogb(block_size-1)+1);
+      }
+
+   public:
+       Allocator() = default;
+       Allocator(const Allocator&) = delete;
+       Allocator& operator=(const Allocator&) = delete;
+
+       ~Allocator() {
+          for (auto& block_stack : block_stacks) {
+             for (auto& block : block_stack) {
+                delete[] block;
+             }
+          }
+       }
+
+       TYPE* allocate(const size_t block_size) {
+          if (block_size == 0) return nullptr;
+
+          const size_t block_id = get_block_id(block_size);
+          if (block_id >= block_stacks.size()) {
+             block_stacks.resize(block_id+1);
+          }
+
+          const size_t allocation_size = 1 << block_id;
+          if (block_stacks[block_id].empty()) {
+             TYPE* block = new TYPE[allocation_size];
+             if (!Array<TYPE>::s_standard_type) {
+                for (size_t k = 0; k < allocation_size; ++k) {
+                   block[k].~TYPE();
+                }
+             }
+             block_stacks[block_id].push_front(block);
+          }
+
+          TYPE* block = block_stacks[block_id].front();
+          block_stacks[block_id].pop_front();
+          if (!Array<TYPE>::s_standard_type) {
+             for (size_t k = 0; k < block_size; k++) {
+                (void) new (&block[k]) TYPE;
+             }
+          }
+          return block;
+       }
+
+       void deallocate(TYPE* const block, const size_t block_size) {
```

Comment:
```suggestion
       void deallocate(TYPE* const block, const size_t used_array_size) {
```
or something similar - it isn't equal to the block size (the code is correct, we just need a more descriptive name)

Reply:
This is an implementation detail that doesn't need to go in the interface.